src/programs.c

```diff
@@ -637,5 +778,1001 @@ int uprobe__read_return_string(struct pt_regs *ctx)
+    if (ret < 0)
+    {
+        return;
+    }
+
+    // Update the index
+    // this is fine:
+    idx = *pcurrent_index + 1;
+    // // This fails verification on 4.15 with `invalid register spill`
+    // idx += 1;
+    // // this fails too, same message
+    // idx = idx + 1;
+    ret = bpf_map_update_elem(&telemetry_index, &key, &idx, BPF_ANY);
+    if (ret < 0 )
+    {
+        return;
+    }
+
+    ptelemetry_event_t pev = bpf_map_lookup_elem(&telemetry_stack, &idx);
+    if (pev)
+    {
+        __builtin_memcpy(ev, pev, sizeof(telemetry_event_t));
+    }
+    return;
+}
+
+#define FLUSH_LOOP                                                          \
+    if (ii < *pcurrent_index) {                                            \
+        ev = bpf_map_lookup_elem(&telemetry_stack, &ii);                     \
+        if (ev) {                                                           \
+            bpf_perf_event_output(ctx,                                      \
+                &telemetry_events,                                          \
+                bpf_get_smp_processor_id(),                                 \
+                ev,                                                         \
+                sizeof(*ev)                                                 \
+            );                                                              \
+            ii++;                                                        \
+        }                                                                   \
+        else                                                                \
+        {                                                                   \
+        }\
+    }
+
+
+static __always_inline void flush_telemetry_events(struct pt_regs *ctx)
+{
+    u32 key = 0;
+    u32 *pcurrent_index = bpf_map_lookup_elem(&telemetry_index, &key);
+    if (NULL == pcurrent_index)
+    {
+        return;
+    }
+
+    u32 *pii = bpf_map_lookup_elem(&read_flush_index, &key);
+    if (NULL == pii)
+    {
+        bpf_map_update_elem(&read_flush_index, &key, &key, BPF_ANY);
+        return;
+    }
+
+    u32 ii = *pii;
+    ptelemetry_event_t ev = NULL;
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    FLUSH_LOOP
+    bpf_map_update_elem(&read_flush_index, &key, &ii, BPF_ANY);
+}
+
+#define READ_LOOP(PTR, T)                                                       \
+    ev->id = id;                                                                \
```

Comment:
Do the ev assignments need to happen every loop unroll? it seems like we are resusing the same memory space

Reply:
That's leftover from when `push_telemetry_event()` would return a fresh event from the map/stack, I'll move these outside the loop so we save a few instructions.

Reply:
750d697f2242e69233ef10fb19251be28b8c7bae

Reply:
Turns out this was a bad idea.